websocket/src/websocket.cpp

```diff
@@ -0,0 +1,536 @@
+            if (RESULT_OK != conn->m_Status)
+            {
+                HandleCallback(conn, EVENT_ERROR);
+            }
+
+            HandleCallback(conn, EVENT_DISCONNECTED);
+
+            g_Websocket.m_Connections.EraseSwap(i);
+            --i;
+            --size;
+            DestroyConnection(conn);
+        }
+        else if (STATE_CONNECTED == conn->m_State)
+        {
+#if defined(HAVE_WSLAY)
+            int r = WSL_Poll(conn->m_Ctx);
+            if (0 != r)
+            {
+                CLOSE_CONN("Websocket closing for %s (%s)", conn->m_Url.m_Hostname, WSL_ResultToString(r));
+                continue;
+            }
+            r = WSL_WantsExit(conn->m_Ctx);
+            if (0 != r)
+            {
+                CLOSE_CONN("Websocket received close event for %s", conn->m_Url.m_Hostname);
+                continue;
+            }
+#else
+            int recv_bytes = 0;
+            dmSocket::Result sr = Receive(conn, conn->m_Buffer, conn->m_BufferCapacity-1, &recv_bytes);
+            if( sr == dmSocket::RESULT_WOULDBLOCK )
+            {
+                continue;
+            }
+
+            if (dmSocket::RESULT_OK == sr)
+            {
+                conn->m_BufferSize += recv_bytes;
+                conn->m_Buffer[conn->m_BufferCapacity-1] = 0;
+                conn->m_HasMessage = 1;
+            }
+            else
+            {
+                CLOSE_CONN("Websocket failed to receive data %s", dmSocket::ResultToString(sr));
+                continue;
+            }
+#endif
+
+            if (conn->m_HasMessage)
+            {
+                HandleCallback(conn, EVENT_MESSAGE);
+                conn->m_HasMessage = 0;
+                conn->m_BufferSize = 0;
+            }
+        }
+        else if (STATE_HANDSHAKE_READ == conn->m_State)
+        {
+            Result result = ReceiveHeaders(conn);
+            if (RESULT_WOULDBLOCK == result)
+            {
+                continue;
+            }
+
+            if (RESULT_OK != result)
+            {
+                CLOSE_CONN("Failed receiving handshake headers. %d", result);
+                continue;
+            }
+
+            result = VerifyHeaders(conn);
+            if (RESULT_OK != result)
+            {
+                CLOSE_CONN("Failed verifying handshake headers:\n%s\n\n", conn->m_Buffer);
+                continue;
+            }
+
+#if defined(HAVE_WSLAY)
+            int r = WSL_Init(&conn->m_Ctx, g_Websocket.m_BufferSize, (void*)conn);
+            if (0 != r)
+            {
+                CLOSE_CONN("Failed initializing wslay: %s", WSL_ResultToString(r));
+                continue;
+            }
+
+            dmSocket::SetNoDelay(conn->m_Socket, true);
+            // Don't go lower than 1000 since some platforms might not have that good precision
+            dmSocket::SetReceiveTimeout(conn->m_Socket, 1000);
+            if (conn->m_SSLSocket)
+                dmSSLSocket::SetReceiveTimeout(conn->m_SSLSocket, 1000);
```

Comment:
Each frame, the timeout is 1ms.
If this becomes an issue, we'll need to put it on another thread.